adapters/tlsio_mbedtls.c

```diff
@@ -631,12 +631,25 @@ int tlsio_mbedtls_close(CONCRETE_IO_HANDLE tls_io, ON_IO_CLOSE_COMPLETE on_io_cl
         }
         else
         {
-            tls_io_instance->tlsio_state = TLSIO_STATE_CLOSING;
-            if (tls_io_instance->tls_status == TLS_STATE_INITIALIZED)
+            if (tls_io_instance->tlsio_state == TLSIO_STATE_ERROR)
```

Comment:
This can be rewritten to simplify the code. What do you think about:

```c
tls_io_instance->tls_status = TLS_STATE_CLOSING; 
if (tls_io_instance->tls_status == TLS_STATE_INITIALIZED) 
{
	if (tls_io_instance->tlsio_state == TLSIO_STATE_ERROR)
	{
		// forced shutdown if tls is in ERROR state
		mbedtls_ssl_session_reset(&tls_io_instance->ssl);
	}
	else
	{
		mbedtls_ssl_close_notify(&tls_io_instance->ssl);
	}
	
	tls_io_instance->tls_status = TLS_STATE_CLOSING; 
} 
```

Reply:
Ack.

---
In reply to: [353445671](https://github.com/Azure/azure-c-shared-utility/pull/419#discussion_r353445671) [](ancestors = 353445671)

Reply:
This will never execute if loop as we are setting to close and then checking init.

how about if I do this:

int is_error_state = tls_io_instance->tlsio_state == TLSIO_STATE_ERROR;
tls_io_instance->tlsio_state = TLSIO_STATE_CLOSING;
if (tls_io_instance->tls_status == TLS_STATE_INITIALIZED)
{
                
                if ( is_error_state)
                {
                   // forced shutdown if tls is in ERROR state
                    mbedtls_ssl_session_reset(&tls_io_instance->ssl);
                }
                else
                {
                    // Tell the peer that you're going to close
                    mbedtls_ssl_close_notify(&tls_io_instance->ssl);
                }

                tls_io_instance->tlsio_state = TLSIO_STATE_CLOSING;
}

---
In reply to: [353446450](https://github.com/Azure/azure-c-shared-utility/pull/419#discussion_r353446450) [](ancestors = 353446450,353445671)